[Relay] Partial Evaluation

MarisaKirisame opened the pull request:
During the pass two weeks, i had [prototyped](https://github.com/MarisaKirisame/PE/blob/master/bin/test.ml) a partial evaluator in metaocaml for a toy DSL with similar characteristic to relay.

It works by taking a evaluator, lifting the value to partially-static domain, reifying the store, and anfing the code generated using let list to avoid code duplication and get capture avoidance substitution for free.

It is written such that it could be translated line by line to tvm.

I had wrote some test, and it can indeed remove the closure and reference #2496 generate.
However, it does generate lots of dead code (closure, reference) that need to be remove by the dead code elimination pass. We probably need an ad hoc naive escape analysis as only write to dead variable is ok to remove.

I will port this but the high level idea is ready for review: the prototype is only 336 loc (including all the DSL definition), and the core function is less then 100 line.

Thanks to @weberlo for reviewing the doc

MarisaKirisame commented:
This is now feature complete. I will add some test and it will be good.
@icemelon9 @masahi @junrushao1994 @FrozenGene @nhynes @wweic  can you guys give a round of review?

nhynes commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,714 @@
+ * There will be a mutable mapping of id to partially static value, called the store.
+ * This allow us to rollback the store:
+ * when a path may or may not get executed, we copy the store, recurse with a copy,
+ * and when we are done, we continue with the old store, so those effect 'do not happened'.
```
perhaps

```
when a path may or may not be executed (as in a conditional), we copy the store,
recursively Reify with the copy, and reinstate the original when the call returns
so that that the effects of the computation are not preserved.
```

nhynes commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,714 @@
+ * as, when we see a function, we partially evaluate it with all the argument as dynamic,
+ * to generate efficient dynamic for that function.
+ *
+ * 3: The code generated will reuse binding (although they are not shadowed),
```
```suggestion
 * 3: The generated code reuses bindings (although they are not shadowed),
```

to follow the verb structure of the previous two numbered summaries

nhynes commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,714 @@
+ * to generate efficient dynamic for that function.
+ *
+ * 3: The code generated will reuse binding (although they are not shadowed),
+ * so we has to dedup them.
```
```suggestion
 * so we have to deduplicate them.
```

nhynes commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,714 @@
+ * Also, It will also generate lots of dead code,
+ * so it is a good idea to feed it through the dead code eliminator after partial evaluation.
+ *
+ * There are some rooms for obvious improvement.
```
```suggestion
 * The partial evaluator makes several assumptions, so there is room for improvement:
```

nhynes commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,714 @@
+ *
+ * There are some rooms for obvious improvement.
+ *
+ * 0: The partial evaluator treat global variable as opaque.
```
```suggestion
 * 0: The partial evaluator treats global variables as opaque.
```

nhynes commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,714 @@
+ * Instead, we can get a match tree, pair it with the data and evaluate it to a normal form.
+ * We then can reify the result.
+ *
+ * Luckily, this issues do not effect the correctness of the algorithm.
```
```suggestion
 * These assumptions do not affect the correctness of the algorithm, however.
```

nhynes commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,714 @@
+ private:
+  std::list<Frame> env_;
+
+  struct RAII {
```
I think that this might be more aptly named `FrameContext` than the generic term `RAII`

nhynes commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,714 @@
+                   const Attrs& attrs,
+                   const tvm::Array<Type>& type_args,
+                   LetList* ll) {
+        return env_.Extend<PStatic>([&]() {
```
I could be mistaken, but it looks like all of the values in scope are references. It'll be more efficient to just copy them using `[=]` instead of taking `&&` with `[&]` (but my guess is that it doesn't actually matter because compilers are good).

MarisaKirisame replied:
I'd like & as it is more 'extensible': if you add a value outside that got used it wont copy it

nhynes commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,714 @@
+        ns_args.push_back(ps->dynamic);
+      }
+      PStatic ns = NoStatic(CallNode::make(expr, ns_args, attrs, type_args));
+      if (!StatefulOp(expr)) {
```
Maybe do an early return and dedent the remaining code.

```cxx
if (StatefulOp(expr))
	return;

tvm::Array<Expr> args;
for (const PStatic& ps : pv) {
// ...
```

nhynes commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,714 @@
+      return TupleNode::make(fields);
+    } else {
+      LOG(ERROR) << "Unknown case";
+      throw;
```
is this error actually recoverable (and not evidence of a bogus program)? if not `LOG(FATAL) << "Unknown case";` might be more appropriate.

nhynes commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,714 @@
+            continue;
+          case MatchStatus::NoMatch:
+            current_match_status = MatchStatus::NoMatch;
+            continue;
```
```suggestion
            break;
```
This is an extremely minor optimization that's possible because the match status will never become anything but `NoMatch` after going `NoMatch`

nhynes commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,714 @@
+            continue;
+          case MatchStatus::Unknown:
+            if (current_match_status == MatchStatus::Match) {
+              current_match_status = MatchStatus::Unknown;
```
```suggestion
              current_match_status = MatchStatus::Unknown;
              break;
```

MarisaKirisame replied:
Why break here?

nhynes replied:
because once `current_match_status` becomes Unknown, it won't change (in fact, every subsequent loop will have the conditional evaluate to `false`)

MarisaKirisame replied:
It can change to NoMatch

tqchen commented on include/tvm/relay/expr.h:
```diff
@@ -166,6 +166,18 @@ class VarNode : public ExprNode {
 
 RELAY_DEFINE_NODE_REF(Var, VarNode, Expr);
 
+struct VarHash {
```
You can simply use NodeHash and NodeEqual without introducing these

MarisaKirisame replied:
NodeHash and NodeEqual is wrong, pointer equality for var is bad as var is based by id. In fact I only realize after using NodeHash and getting error.

tqchen replied:
OK, please add more comment on they check equivalence by id.

tqchen commented:
tagging more reviewers, @wweic @junrushao1994

junrushao1994 commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,719 @@
+ * In the extreme case, partial evaluation can even turn the whole program
+ * into pure first order computation with no control.
+ * In such a case, we can compile the whole computation onto SIMD Instruction/GPU/FPGA,
+ * and get insane speedup.
```
Remove insane :-)

junrushao1994 commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,719 @@
+ * This avoids code duplication (which is both inefficient and incorrect), as atom has constant size
+ * and allow us to not handle capture-avoidance substitution (as atom has no binder).
+ *
+ * 2: The map of References to partially static values is Reified, as described below.
```
lowercase Reified?

junrushao1994 commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,719 @@
+ * There will be a mutable mapping of id to partially static value, called the store.
+ * This allow us to rollback the store:
+ * when a path may or may not be executed (as in a conditional), we copy the store,
+ * recursively Reify with the copy, and reinstate the original when the call returns
```
lowercase Reify?

junrushao1994 commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,719 @@
+
+using namespace runtime;
+
+struct StaticNode {
```
Why not just inherit the TVM's node system? Should be similar i guess

MarisaKirisame replied:
It is considerably more work, and we has no need.

tqchen replied:
While it might take a bit more work to adopt TVM's node system, it makes us share the same infra and reuse later. I will recommend using node instead here.

Note that it is not too much more work, just replace shared_ptr->NodePtr and make_shared->make_node

tqchen commented on include/tvm/relay/expr.h:
```diff
@@ -166,12 +166,20 @@ class VarNode : public ExprNode {
+ * Different VarNode might has same vid, and they are considered to be the same var in such case.
+ * Use VarHash to hash Var by id.
+ */
 struct VarHash {
```
On second thought, while I know that Var can be defined by using identity in id, it is dangerous to make this assumption. Most of the current code will keep a unique Var during transformation(and transform all occurrences of Var if its type signature change) So likely we don't need to rely on the comparison by id within a pass. Please check if the pass can be improved.

MarisaKirisame replied:
No It cannot. Typesubst get called multiple time. I can write a pass that bind them back to a single one, but I found out I dont need to for our test case right now, as dce is working fine.

wweic commented:
Nice work. I'll review today. Interesting to see its effect in the future. As I understand, majority of relay function calls are ADT utilities and recursion generated by framework converter. I don't think PE can help the latter case since it's conditioned on dynamic value, not easy to handle by PE.

MarisaKirisame commented:
@wweic we are planning to use PE to remove the closure/reference of higher order ad as much as possible.
additionally, it help on ADT utilities, as it can do some fusion. for example, I think this can do map f . map g = map (f . g) once it get a termination checker.

wweic commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,764 @@
+              return VisitExpr(op->false_branch, ll)->dynamic;
+            });
+        });
+      store_.Invalidate();
```
Is this `Invalidate` call needed if you are creating a new scope for true/false branch and exit it afterwards via `store.Extend`?

MarisaKirisame replied:
Yes.
suppose that both if and else branch are the same, and they contain a write to an unknown reference.

wweic replied:
Thanks. Could you add a test case? I disabled `Invalidate` and run `test_pass_partial_eval.py`, `test_pass_dead_code_elimination.py`, no test failure. I think it's good to have the test coverage.

MarisaKirisame replied:
sure

wweic commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,764 @@
+  return DeDupMutator().VisitExpr(e);
+}
+
+Expr Remap(const Expr& e) {
```
Could you add a comment about what it does? What is it remapping?

wweic commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,764 @@
+  return TypeVarNode::make(tv->var->name_hint, tv->kind);
+}
+
+Expr DeDup(const Expr& e) {
```
Could you add some comment about this pass?

MarisaKirisame commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,764 @@
+ * In such a case, we can compile the whole computation onto SIMD Instruction/GPU/FPGA,
+ * and get huge speedup.
+ *
+ * It works by making the following modifications to the standard relay interpreter:
```
The code is very similar to the relay interpreter. By modification I mean conceptually, you start with the interpreter, modify some part of it, and get this file.

wweic replied:
Ah, I was under the impression that the pass is only for interpreter. Seems like it's not the case.

MarisaKirisame commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,764 @@
+  return Static(make_node<SRefNode>());
+}
+
+using Clos = std::function<PStatic(const std::vector<PStatic>&,
```
Closure is used, plus this isnt really Closure, so I changed it to Func.

wweic commented:
@MarisaKirisame I'm looking at the result of partial evaluator of `test_if_ref`:
````python
In [16]: print(f.astext())
v0.0.1
%13 = fn (%d: bool) -> int32 {
  %0 = ref(1)
  %12 = {
    let %r: ref(int32) = %0
    %5 = fn () -> () {
      %1 = %r^
      %2 = %r^
      %3 = add(%1, %2) // ty=int32
      %4 = (%r := %3)
      %4
    }
    %11 = {
      let %u: fn () -> () = %5
      %8 = if (%d) {
        %6 = %u() // ty=()
        %6
      } else {
        %7 = %u() // ty=()
        %7
      }
      %10 = {
        let %eff: () = %8
        %9 = %r^
        %9
      }
      %10
    }
    %11
  }
  %12
}
%13

In [17]: print(pe_f.astext())
v0.0.1
%18 = fn (%d: bool) -> int32 {
  %17 = {
    let %x: int32 = 1
    %0 = ref(%x)
    %16 = {
      let %x1: ref(int32) = %0
      %4 = fn () -> () {
        %3 = {
          let %x3: int32 = 2
          %1 = (%x1 := %x3)
          %2 = {
            let %x4: () = %1
            %x4
          }
          %2
        }
        %3
      }
      %15 = {
        let %x2: fn () -> () = %4
        %11 = if (%d) {
          %7 = {
            let %x6: int32 = 2
            %5 = (%x1 := %x6)
            %6 = {
              let %x7: () = %5
              %x7
            }
            %6
          }
          %7
        } else {
          %10 = {
            let %x8: int32 = 2
            %8 = (%x1 := %x8)
            %9 = {
              let %x9: () = %8
              %x9
            }
            %9
          }
          %10
        }
        %14 = {
          let %x5: () = %11
          %12 = %x1^
          %13 = {
            let %x10: int32 = %12
            %x10
          }
          %13
        }
        %14
      }
      %15
    }
    %16
  }
  %17
}
%18
````

I'm not sure if partial evaluator is safe to inline the `update` closure. `update` reads from a ref in the beginning, so it depends on the ref. I don't think we can simply propagate the constant `1` into the closure, right? `update` closure captures the reference, but not the value. Does it make sense?

MarisaKirisame commented:
@junrushao1994 can you approve?

wweic commented:
@MarisaKirisame I modified your test case a bit, instead of creating a ref from const 1, I pass in an argument and create ref from there. I think PE did an incorrect inlining:
````python
In [5]: print(pe_f.astext())
v0.0.1
%17 = fn (%d: bool, %i: int32) -> int32 {
  %0 = ref(%i)
  %16 = {
    let %x: ref(int32) = %0
    %4 = fn () -> () {
      %1 = add(%i, %i) // ty=int32
      %2 = (%x := %1)
      %3 = {
        let %x2: () = %2
        %x2
      }
      %3
    }
    %15 = {
      let %x1: fn () -> () = %4
      %11 = if (%d) {
        %5 = add(%i, %i) // ty=int32
        %6 = (%x := %5)
        %7 = {
          let %x4: () = %6
          %x4
        }
        %7
      } else {
        %8 = add(%i, %i) // ty=int32
        %9 = (%x := %8)
        %10 = {
          let %x5: () = %9
          %x5
        }
        %10
      }
      %14 = {
        let %x3: () = %11
        %12 = %x^
        %13 = {
          let %x6: int32 = %12
          %x6
        }
        %13
      }
      %14
    }
    %15
  }
  %16
}
%17

In [6]: print(f.astext())
v0.0.1
%13 = fn (%d: bool, %i: int32) -> int32 {
  %0 = ref(%i)
  %12 = {
    let %r: ref(int32) = %0
    %5 = fn () -> () {
      %1 = %r^
      %2 = %r^
      %3 = add(%1, %2) // ty=int32
      %4 = (%r := %3)
      %4
    }
    %11 = {
      let %u: fn () -> () = %5
      %8 = if (%d) {
        %6 = %u() // ty=()
        %6
      } else {
        %7 = %u() // ty=()
        %7
      }
      %10 = {
        let %eff: () = %8
        %9 = %r^
        %9
      }
      %10
    }
    %11
  }
  %12
}
%13
````

It's aggressively inlines the value into the closure, but I think closure only captures the reference, not the value, right?

MarisaKirisame commented:
@wweic good catch, that is indeed incorrect. can you give me the offending test? I know how to fix it but wanna add it to the test case.

wweic commented:
@MarisaKirisame 
````python
def test_if_ref():
    shape = ()
    dtype = 'bool'
    t = relay.TensorType(shape, dtype)
    d = relay.Var("d", t)
    i = relay.Var("i", relay.TensorType(shape, 'int32'))    
    r = relay.Var("r")
    update = relay.Function([], relay.RefWrite(r, relay.RefRead(r) + relay.RefRead(r)))
    u = relay.Var("u")
    body = relay.If(d, u(), u())
    eff = relay.Var("eff")
    body = relay.Let(eff, body, relay.RefRead(r))
    f = relay.Function([d, i], relay.Let(r, relay.RefCreate(i), relay.Let(u, update, body)))
    f = infer_type(f)
    pe_f = infer_type(partial_eval(f))
````

MarisaKirisame commented:
@wweic I had updated with a regression test to the error.

wweic commented on src/relay/pass/partial_eval.cc:
```diff
@@ -0,0 +1,766 @@
+    return HasStatic(MkSFunc(f), GetRef<Expr>(op));
+  }
+
+  PStatic VisitExpr_(const MatchNode* op, LetList* ll) final {
```
Could you add a test case for pattern match? For example i'd like to see the following test case(really powerful):
Call `hd`  on `cons(1, cons(3, nil))`, it gets inlined to match expression, then simplified to 1.

MarisaKirisame replied:
It doesnt support module yet. I can define my own head and call it instead of using prelude, is it fine with you? We can get module next time as it is ton of work and there is too much in this PR

wweic replied:
Oh, I'm fine with it. I just want to see adt + function inline + pattern match interplay together.

MarisaKirisame commented:
@wweic the test is added.

MarisaKirisame commented:
@tqchen @junrushao1994 @wweic I addressed all the review feedback, can you guys approve explicitly? (https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly)

MarisaKirisame commented:
@tqchen this is good. can you give some review?

tqchen commented on python/tvm/relay/ir_pass.py:
```diff
@@ -937,3 +937,20 @@ def pass_debug_print(ast, show_meta_data=True, annotate=None, gnf=True):
     return _ir_pass.pass_debug_print(ast, show_meta_data, annotate, gnf)
+
+
+def partial_eval(expr):
```
partial_evaluate

tqchen commented on include/tvm/relay/expr.h:
```diff
@@ -166,6 +166,26 @@ class VarNode : public ExprNode {
+ * Different VarNode might has same vid, and they are considered to be the same var in such case.
+ * Use VarHash to hash Var by id.
+ */
+struct VarHash {
```
consider move this to internal header then, since it is used within the pass

MarisaKirisame commented:
@tqchen I address your comment. can you merge?
